### ISSUE TITLE
Pyright type inference failing with function that returns an awaitable class

### ISSUE THREAD

**Issue by dob9601** [microsoft/pyright#1336]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
`pyright --lib` infers incorrect type hints for the `sync_to_async` method of asgiref. It seems to be struggling with the fact that the `SyncToAsync` class that `sync_to_async` returns has an asynchronous `__call__` function.

**To Reproduce**
```py
from asgiref.sync import sync_to_async

def foo():
    return 'bar'

async def test():
    await sync_to_async(foo)([1, 2, 3])

```
and then call `pyright --lib` on the file. Several different warnings are generated, which are:
- Argument missing for parameter "f"
- No parameter named "x" (where x is a parameter of the synchronous function you passed)
- "SyncToAsync" is not awaitable (which is the awaitable function that the sync_to_async method returns)

**Expected behavior**
No type warnings to be raised

**Screenshots or Code**
Sample to reproduce provided in the "To Reproduce" section.

**VS Code extension or command-line**
coc-pyright - but also tested and found to occur with the command-line version of pyright when the `--lib` flag is used

**Additional context**
I'm aware that I can create stubs for the method, just thought it would be worth raising that it seemed to be causing issues for pyright

**Comment by erictraut** [comment]
I think Pyright is doing the right thing here. It is inferring that the return type of `sync_to_async` is `Union[Callable[Unknown, SyncToAsync], SyncToAsync]`, which is correct. Both of the types in this union are callable, but they can't be called with the same parameter lists, and they return different types (SyncToAsync and a Coroutine, respectively). The first of these types is not awaitable, and the second is. So all of the error messages look correct to me.

For a function with such an odd signature, the correct way to annotate it is with an overload that specifies different return results based on the `func` parameter type.

**Comment by dob9601** [comment]
Noted, thanks. Is there any better way than creating overloads? I've had to wrap loads of different Django methods in `sync_to_async` so it'd result in quite the long type file.

**Comment by erictraut** [comment]
For a poorly-conceived function like `sync_to_async` that returns completely different types based on its inputs, overloads are your only option if you want to call it directly.

If you don't mind creating helper functions within your code that "wrap" sync_to_async, you could create two wrappers, one that accepts an input func and one that doesn't. You could then annotate these helper functions to eliminate the need for any type inference. That's how I'd address this in my own code.

**Comment by dob9601** [comment]
Great idea, I'll give that a go :)
Thanks very much for your help. I get the impression from the Django docs that sync_to_async is still very much a temporary solution until it has full async support
